app/controllers/concerns/two_factor_authenticatable_methods.rb

```diff
@@ -124,6 +127,22 @@ def handle_remember_device_preference(remember_device_preference)
     save_remember_device_preference(remember_device_preference)
   end
 
+  def increment_mfa_selection_attempt_count(auth_method)
+    user_session[:mfa_attempts] ||= {}
+    user_session[:mfa_attempts][:attempts] ||= 0
+    if user_session[:mfa_attempts][:auth_method] != auth_method
+      user_session[:mfa_attempts][:attempts] = 0
+    end
+    user_session[:mfa_attempts][:attempts] += 1
+    user_session[:mfa_attempts][:auth_method] = auth_method
+  end
+
+  def mfa_attempts_count
+    if user_session[:mfa_attempts]
+      user_session[:mfa_attempts][:attempts]
+    end
```

Comment:
We could use [`Hash#dig`](https://www.rubydoc.info/stdlib/core/Hash:dig) to simplify this and gracefully handle `nil` `mfa_attempts`. Arguably this could be inlined everywhere to avoid creating a helper method, though I think it can also be helpful to abstract access to user session properties.

```suggestion
    user_session.dig(:mfa_attempts, :attempts)
```

Reply:
That's much nicer. I agree with keeping it abstracted.